dwave/optimization/include/dwave-optimization/c++23/cartesian_product.hpp

```diff
@@ -0,0 +1,23 @@
+// Copyright 2023 D-Wave Systems Inc.
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+#pragma once
+
+#include <vector>
+
+namespace dwave::optimization {
+
+    std::vector<std::vector<ssize_t>> cartesian_product(std::vector<std::vector<ssize_t>> v);
```

Comment:
I'd put this in [utils.hpp](https://github.com/dwavesystems/dwave-optimization/blob/main/dwave/optimization/include/dwave-optimization/utils.hpp) rather than a new folder. And just note that it's a backport. [Like we do with `unreachable()`](https://github.com/dwavesystems/dwave-optimization/blob/d27a6385925c9fde5fddf011750da28a2b873e8e/dwave/optimization/include/dwave-optimization/utils.hpp#L32).